### ISSUE TITLE
Add AAAA Record to Sandbox Domain (IPv6)

### ISSUE THREAD

**Issue by motters** [eclipse/leshan#1205]
It would be super useful for testing / eval if the sandbox server domain (https://leshan.eclipseprojects.io/) had an IPv6 address (AAAA record).

For example, when using OpenThread for connectivity it's an IPv6 network only. Conversions are possible but it's not as good as keeping IPv6 throughout.

**Comment by sbernard31** [comment]
By the past, I remembered that the sandbox server was IPv4 only.
I'm not sure if this is still true, I can double check this with eclipse webmaster.

Waiting you can run leshan-server-demo in your own IPv6 environment. (Leshan should support IPv6)

**Comment by sbernard31** [comment]
(https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/904)

**Comment by motters** [comment]
Cheers Simon for checking this out.

I've set up Leshan on a VM (https://hub.docker.com/r/corfr/leshan) temporarily.

**Comment by motters** [comment]
@sbernard31 It certainly does now have an IPv6 however i'm failing to connect to the bootstrap server. This could be my end  as i'm trying to get DTLS working (openthread device lwm2m -> openthread border router -> Leshan).

I'll be looking at it today.

**Comment by boaks** [comment]
If you run your own sandbox, could you please provide [ip-capture](https://github.com/eclipse/californium/wiki/Logs-and-IP-Capturing-%E2%80%90-How-To-Provide-The-Right-Information#ip-capturing)?

**Comment by motters** [comment]
@boaks I'm using Leshan demo server atm, however i'll try it with my own sand box today as well.

**Comment by boaks** [comment]
What's "atm"?

**Comment by motters** [comment]
@boaks My bad, "at the moment". I'll test my sandbox today though.

**Comment by boaks** [comment]
One of the common issues with IPv6 is handling multiple IPv6 network interfaces, see [local wildcard address - ipv6](https://github.com/eclipse/californium/issues/315). Not sure, what the leshan demo server does.

**Comment by boaks** [comment]
I don't get a response from the leshan sandbox as well. Neither coap not dtls. 

@sbernard31 
Does the demo-server use the "local wildcard" address?

**Comment by motters** [comment]
No packets are found on my sandbox
`sudo tcpdump -s 0 -w temp.pcap -i any '(udp port 5683) or (udp port 5684) or (udp port 53) or (udp port 5784)'`

I shall keep looking.

**Comment by boaks** [comment]
Hm ... on a private sandbox I get IPv6 traffic.
for some clouds the process to set it up seems to be a little more complex.

[Californium - Cloud Setup](https://github.com/eclipse/californium/wiki/Californium-as-old-style-unix-systemd-service#cloud-init---automatic-cloud-vm-installation)

Especially for "hyper scaler" the setup was more complicated, so I skipped it for my instructions.

**Comment by sbernard31** [comment]
>  Does the demo-server use the "local wildcard" address?

Yes it currently uses wildcard address.

I quickly looked at [local wildcard address - ipv6](https://github.com/eclipse/californium/issues/315) and I probably missed a lot about this problem. 
Currently, I understand the issue is when several IPv6 address is assigned,  correct ? 
Does it means if we have only 1 IPv4 and 1 IPv6 addresses this should works  with local wildcard ? 

I also understand that if we have several IPv6 addresses with 1IPv4 address, the only way to make it works (at java level)  is to use 1 connector with IPv4 and another connector with Ipv6 (with static binding, I mean not local wildcard) ?

**Comment by sbernard31** [comment]
Current sandbox IP : 
```
>> ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:0d:3a:38:84:54 brd ff:ff:ff:ff:ff:ff
    inet 10.0.10.9/24 brd 10.0.10.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 ace:ceb:deca:deed::4/128 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::20d:3aff:fe38:8454/64 scope link 
       valid_lft forever preferred_lft forever
```

**Comment by boaks** [comment]
Unfortunately, I'm not too common with that. The Californium Plugtest server start one connector for each address and that works quite well. If other, better solutions are possible, is unknown to me.

Beside of that issue, it's also not that easy to setup ipv6 proper in some clouds. As I wrote above, I was only able to configure it for some clouds, in others it seems to be too complicated for me ;-).

**Comment by boaks** [comment]
I request the [ipv6 setup for californium.eclipseprojects.io](https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/917)  as well. When it's available, we see, if its the californium setup or the vm setup.

**Comment by motters** [comment]
@boaks I'm using digital ocean so will take a look at your scripts now.

**Comment by boaks** [comment]
Should I prepare a temporary Californium ipv6 sandbox on DO for 1-2 days? I don't want to assign a dns address, but if a literal ipv6 will do it, that can be up in a couple of minutes.

**Comment by boaks** [comment]
I setup a Californium Plugtest Server with IPv6 on Exoscale.

URL: coaps://[2a04:c46:e00:979c:489:28ff:fe00:182]:5684/myip

I started a capture there. You may try a handshake using the openssl default PSK credentials,
"Client_identity" and  "secretPSK".

It's no lwm2m, you mainly will see, if you get some messages back.

**Comment by sbernard31** [comment]
> The Californium Plugtest server starts one connector for each address and that works quite well. If other, better solutions are possible, is unknown to me.

Unfortunately leshan-server-demo supports only 2 connectors :  one for UDP and one for DTLS.
I don't plan to change this at short term.

**Comment by boaks** [comment]
sure. do so.

**Comment by boaks** [comment]
> Unfortunately leshan-server-demo supports only 2 connectors : one for UDP and one for DTLS.
> I don't plan to change this at short term.

AFAIK, some client may then chose the wrong one on a dns lookup.

```
nslookup leshan.eclipseprojects.io
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
Name:	leshan.eclipseprojects.io
Address: 23.97.187.154
Name:	leshan.eclipseprojects.io
Address: 2603:1020:201:f::60
```

**Comment by boaks** [comment]
> 13:36:10.152674 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
> 13:36:10.785239 IP6 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825 > cali.5684: UDP, length 55
> 13:36:10.786548 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
> 13:36:11.265133 IP6 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825 > cali.5684: UDP, length 55
> 13:36:11.266784 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
> 13:36:11.609443 IP6 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825 > cali.5684: UDP, length 55
> 13:36:11.610792 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
>

**Comment by boaks** [comment]
@motters 

At least your peer sends some messages and gets some response.

Not sure, what's the best to continue.

Do you like to modify the leshan demo server? I could assist.

Alternatively, you may try out [cf-nat](https://github.com/eclipse/californium/tree/master/cf-utils/cf-nat). Start that with the list of your DO vm's ipv6 addresse(s) and the leshan ipv4 destination. Usually it should work, if you don't use the wildcard but explicit the ipv6 address.

**Comment by motters** [comment]
> > 13:36:10.152674 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
> > 13:36:10.785239 IP6 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825 > cali.5684: UDP, length 55
> > 13:36:10.786548 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
> > 13:36:11.265133 IP6 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825 > cali.5684: UDP, length 55
> > 13:36:11.266784 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
> > 13:36:11.609443 IP6 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825 > cali.5684: UDP, length 55
> > 13:36:11.610792 IP6 cali.5684 > 2a01:4b00:e027:5b00:55e0:78fe:4116:a0e6.62825: UDP, length 98
@boaks 

That is my IP address. I successfully connected to your server using your coaps client and got a reply. 

I'll now set up my device to try and access it through lwm2m, see if I get any response.

I do have a NAT I can use on the openthread border router prefixing "64:ff9b" but we wanted to use straight IPv6.

Do we have an idea of where the Leshan demo server has a problem with IPv6?

**Comment by boaks** [comment]
Without californium's sandbox configured to ipv6, I can only guess, that the "local wildcard address" (for ipv4 and all ipv6 addresses) doesn't work, as described in [local wildcard address - ipv6](https://github.com/eclipse/californium/issues/315).
Or the network of the sandbox isn't setup proper. But that we will know, once my ticket gets processed. I guess next week.

**Comment by boaks** [comment]
@sbernard31 

is it possible to start the demo-server on a specific network interface? If so, motters may just use the ipv6 address of his vm.

**Comment by motters** [comment]
@boaks I'll have a read on "local wildcard address" so I understand more.  But happy to look into modifying the demo server if needed. 

I've sent some packets to your server via the device but don't get a response. Which probably makes sense as it's LwM2M stuff.
![image](https://user-images.githubusercontent.com/2155687/154695980-f56575f2-a83a-43b3-9ae8-ce350ddac6dd.png)

**Comment by boaks** [comment]
I don't get. Why pictures! that's so hard to see the important info!

Di you send a udp message either to port 5683 or 5684? Then I don't see it in my capture. i fyou send it to 5784, then that isn't shown in my capture.

(In the Californium issue the other user started leshan by using the ipv6 adddress explicit. Not sure, if this is possible with the demo server.)

**Comment by boaks** [comment]
> I've sent some packets to your server via the device but don't get a response. Which probably makes sense as it's LwM2M stuff.

I expect, that if port 5683 or 5684 is used, some messages are exchanged. For 5684 the dtls handshake messages, and for 5683 a error response.

**Comment by boaks** [comment]
Let me add: if the leshan on you DO vm listens on the ports, you should at least see the incoming messages. Youe may also see some outgoing messages. If these messages are using the same local ipv6 address of your vm, it should work. If you see different messages, then not. using leshan or californium doesn't matter. both listen on udp, 5683/5684 and will exchange dtls and coap messages. Just these coap messages are different, though Californium will response with an error, leshan will do the job.

**Comment by motters** [comment]
@boaks It was sent to port 5684 I've attached the wireshake file if that help

That's a good point. I'm seeing nothing at the server let me double check my end again.

[temp server.zip](https://github.com/eclipse/leshan/files/8097870/temp.server.zip)

**Comment by boaks** [comment]
OK, my ip-capture shows no traffic, I guess, there is then something wrong "on your side".

**Comment by sbernard31** [comment]
I'm bit confused by all those inputs :sweat_smile: , so I tried to investigate this step by step.

Here you can find my first research : https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/904#note_543647

**Comment by sbernard31** [comment]
`2a01:e0a:3a8:d8c0:8809:cda:9245:43ad` it's me, I tried to connect using a leshan client from my local machine.

**Comment by boaks** [comment]
not sure, I only log the addresses, not the traffic. should I really capture it and see, whats in the messages ?

**Comment by boaks** [comment]
I still see traffic incoming, but no outgoing one.

**Comment by sbernard31** [comment]
When I try to capture the traffic on my side, I see only outgoing message as `Continuation Data` (I don't know why the CLIENT_HELLO is not seen as a CLIENT_HELLO)

**Comment by boaks** [comment]
s forgotten in the addres?

**Comment by boaks** [comment]
> 15:20:57.712752 IP6 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103 > cali.5684: UDP, length 107
> 15:20:57.716844 IP6 cali.5684 > 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103: UDP, length 60
> 15:20:57.768549 IP6 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103 > cali.5684: UDP, length 139
> 15:20:57.772437 IP6 cali.5684 > 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103: UDP, length 228
> 15:20:57.901683 IP6 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103 > cali.5684: UDP, length 162
> 15:20:59.902226 IP6 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103 > cali.5684: UDP, length 162
> 15:21:03.905034 IP6 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103 > cali.5684: UDP, length 162
> 15:21:11.905640 IP6 2a01:e0a:3a8:d8c0:8809:cda:9245:43ad.44103 > cali.5684: UDP, length 162

Maybe you chose the right secret ...

**Comment by sbernard31** [comment]
Do you want to do more test ? if yes better to use a chat.
like : https://web.libera.chat/  (#ipv6_leshan)

**Comment by boaks** [comment]
Great!
AFAIK, it doesn't matter, if you use the leshan demo-server or the californium plugtest server, you should see at least incoming traffic.
If it's the right one (coap 5683 / dtls 5684) you will also see some outgoing traffic. If that uses then the right or wrong vm local address, may make the difference. But as first step, we need to see the traffic in the vm.

**Comment by boaks** [comment]
I guess, @motters is better of, if he tries to use his DO sandbox. At least a ip-capture of incoming traffic should work, and we will see, what the outgoing contains (no pictures, captures please :-) ).

**Comment by boaks** [comment]
I have to leave for 1.5h. Back later.

**Comment by sbernard31** [comment]
I tried to capture again traffic on leshan sandbox and didn't see any incoming message, so I guess we need to wait for some inputs from eclipse webmaster.

**Comment by motters** [comment]
@boaks @boaks  I looking into it at the moment. The problem is i'm not seeing any traffic inbound or out at the moment using my device (Zephyr).

Is there a LwM2M test program we can use?

**Comment by motters** [comment]
@boaks Didn't get time at the weekend but will look into it today.

**Comment by boaks** [comment]
Let me propose the following test sequence:

client cf-browser:

1. test POST "coaps://[2a03:b0c0:3:d0:0:0:1260:7001]/rd" . Should result in CoAP error response.
        That verifies, that the client host is able to send ipv6.
        That ipv6 messages are sent back
2. test POST "coaps://[your-ipv6]/rd"
        If you get a CoAP error response, all is fine.
        At least you should be able to capture some incoming traffic, otherwise check your ipv6 setup and firewall.
        Note ipv6 uses different firewall rules.
        If you see outgoing traffic, check the vm-local addresses.
if that works, then use your client, and repeat the step 2 sending messages to your vm.

I hope, that helps. If you have any trouble, let me know.

**Comment by motters** [comment]
@boaks i seem to get traffic at the vm when posting to coaps://[your-ipv6]/rd

[motters-post-rd.zip](https://github.com/eclipse/leshan/files/8108437/motters-post-rd.zip)

Test program output:
![image](https://user-images.githubusercontent.com/2155687/154935888-84a13f1e-da8b-4d81-a24f-3a91fe0a69cc.png)

**Comment by boaks** [comment]
FMPOV, the capture seems to show, that the outgoing traffic is blocked on your vm.
Both IPv4  and IPv6 CLIENT_HELLOs never shows the cookie, so the HelloVerifyRequest seems not to reach them.

**Comment by motters** [comment]
@boaks I can confirm that when using ipv6 via a NAT, hence ipv4 at the server I get packets. 

I however I never get any packets when using IPv6. I can't see anything wrong with the firewall, will looking into it more.
[nat-motters.zip](https://github.com/eclipse/leshan/files/8109366/nat-motters.zip)

**Comment by boaks** [comment]
If "motters-post-rd.zip" is from your vm, the vm receives ipv6.
So I guess, it's something in your open threat setup.
AFAIK, the google group "openthread-users@googlegroups.com" is a good place to get support for that.
Or the manufacturer of your open threat equipment.

**Comment by boaks** [comment]
And just to say: I still think, that

motters-post-rd.zip

shows, that "outgoing" isn't working. It shows incoming traffic from 2a01:4b00:e027:5b00:1e0:915c:5ad0:4572, and outgoing to that, but the messages indicates for me, that the outgoing messages don't reach 2a01:4b00:e027:5b00:1e0:915c:5ad0:4572.

On my DO vm, I see also incoming handshakes from 2a01:4b00:e027:5b00:1e0:915c:5ad0:4572. And there are also outgoing messages. But the messages in mv vm shows, that the outgoing stuff reaches 2a01:4b00:e027:5b00:1e0:915c:5ad0:4572. The handshake then timesout/fails, what is usually caused by wrong secrets. Using "secretPSK" (hex 73656372657450534B) should succeed. Anyway, here the udp-ipv6 message exchange works, that's what a first step should also work on your server.

**Comment by motters** [comment]
@boaks Possible, however today to test the theory that it's an issue with "open thread", I got SNTP working over IPv6. Hence the thread network is passing through IPv6 without issue. It could still be something to do with the Zephyr LwM2M library over openthread/IPv6 I supposed. However i so see the packet sent on WLAN0 but never received at the server. 

My question would be where are we up to in getting the Leshan demo server working with IPv6?

**Comment by sbernard31** [comment]
> My question would be where are we up to in getting the Leshan demo server working with IPv6?

For now we are waiting from eclipse webmaster support to understand why we don't see any incoming Ipv6 traffic on the sandbox machine, (See https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/904#note_543647).

**Comment by boaks** [comment]
> My question would be where are we up to in getting the Leshan demo server working with IPv6?

The leshan sandbox? Still "ongoing".

> It could still be something to do with the Zephyr LwM2M library over openthread/IPv6 I supposed.

Let me say:
I'm confused about your state. Mainly because I tried to describe a process, but the information I get is somehow not related to the process I described.

Using cf-browser ipv6 with your vm, what's the result?

And when the Zephyr LwM2M starts, which port does it use? 5784? 
Maybe you can change that, just to see the handshake with my DO vm to 5684.

**Comment by sbernard31** [comment]
I created a dedicated issue about adding Ipv6 address to Leshan sandbox : #1210.
Looking at last comment at : https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/904, I'm not so sure this will be available soon.

You can still use this issue (#1205) to discuss about @motters specific issue but please keep #1210 focus only on the Ipv6 sandbox task.
@motters eventually you can rename this issue with something more relative to your current issue.  :pray:

**Comment by sbernard31** [comment]
@motters, the eclipse webmaster finally succeed to make it works.
I just tested right now and it works for me now. (See #1210)

**Comment by boaks** [comment]
@motters 

Does IPv6 work for you, now that the sandbox supports it?

**Comment by motters** [comment]
@boaks Sorry got pulled off this for a few days. I'm now back on it as of now. I'll check my setup to the sandbox today.

**Comment by boaks** [comment]
Just to mention:
 [local wildcard address - ipv6](https://github.com/eclipse/californium/issues/315) seems not to apply to the "simple vm", so no special endpoint-setup was required for now.
Very interested in your results, looking forward.

**Comment by sbernard31** [comment]
@motters, Should we close this issue  ?

**Comment by motters** [comment]
@sbernard31 We're rushing for the end of a grant atm, currently i'm just using a NAT. 

I'm 90% sure this is an OpenThread config issue not forwarding the IPv6 through.

Might be worth closing this and I'll create a new ticket when we get round to fixing it (if it's a Leshan issue).

**Comment by sbernard31** [comment]
That sounds good to me :)
I close the issue.
